### ISSUE TITLE
Weight decay

### ISSUE THREAD

**Issue by francoishernandez** [cybertronai/pytorch-lamb#3]
Hi @8enmann ,
Thanks for this repo and the blog post!
I'm working on adding this to [OpenNMT-py](https://github.com/OpenNMT/OpenNMT-py).
Are you confident in the implementation of `weight_decay`? It seems like you kept the 'original' implementation from `torch.optim.Adam`, adding `decay * params` to the gradients, but the equations in the paper are adding this `decay * params` to the calculated update (which you call `adam_step`, used as basis for the update, as well as its norm used for the denominator of the trust ratio.
Using your 'raw' implementation leads to very slow / no learning at all, while changing it to `adam_step += decay * params` seems to have the expected effect (not sure of those results yet, some tuning is still going on).
What do you think?
Thanks!
François

Note: we are trying to tune this on some transformer models for translation task, so some behaviours might be quite different than your task.

**Comment by 8enmann** [comment]
Thanks for the catch! I found that weight decay with the current implementation didn't work well at all on Transformer XL. I'm guessing this is the reason. I'm working on a fix now.
The paper was also updated and clarified a lot. 
<img width="467" alt="Screen Shot 2019-06-15 at 3 25 05 PM" src="https://user-images.githubusercontent.com/1021104/59556962-01c94980-8f82-11e9-98b4-c7e17233dc96.png">
